hbase-mapreduce/src/main/java/org/apache/hadoop/hbase/mob/mapreduce/MobRefReporter.java

```diff
@@ -0,0 +1,485 @@
+                + "is missing but the referenced file is still in the table from that tag, then "
+                + "lookups of these impacted rows would still work.");
+            context.write(MISSING, encodeRows(context, key, rows));
+          }
+        } catch (IOException e) {
+          LOG.error("exception while checking mob area of our table for hfilelinks the point to {}",
+              file, e);
+          context.write(MISSING_IO, encodeRows(context, key, rows));
+        }
+      }
+    }
+
+    /**
+     * reuses the passed Text key. appends the configured seperator and then a comma seperated list
+     * of base64 encoded row keys
+     */
+    private Text encodeRows(Context context, Text key, Iterable<ImmutableBytesWritable> rows) throws IOException {
+      StringBuilder sb = new StringBuilder(key.toString());
+      sb.append(seperator);
+      boolean moreThanOne = false;
+      long count = 0;
+      for (ImmutableBytesWritable row : rows) {
+        if (moreThanOne) {
+          sb.append(",");
+        }
+        sb.append(base64.encodeToString(row.copyBytes()));
+        moreThanOne = true;
+        count++;
+      }
+      context.getCounter("PROBLEM", "MOB_FILES").increment(1L);
+      context.getCounter("PROBLEM", "IMPACTED_ROWS").increment(count);
+      context.getCounter("PROBLEM_ROWS_PER_FILE_DIGITS", Integer.toString(Long.toString(count).length())).increment(1L);
+      key.set(sb.toString());
+      return key;
+    }
+  }
+
+  /**
+   * Main method for the tool.
+   * @return 0 if success, 1 for bad args. 2 if job aborted with an exception,
+   *   3 if unable to start due to other compaction, 4 if mr job was unsuccessful
+   */
+  public int run(String[] args) throws IOException, InterruptedException, ServiceException {
+    // TODO make family and table optional
+    if (args.length != 3) {
+      printUsage();
+      return 1;
+    }
+    final String output = args[0];
+    final String tableName = args[1];
+    final String familyName = args[2];
+    final long reportStartTime = EnvironmentEdgeManager.currentTime();
+    Configuration conf = getConf();
+    try {
+      FileSystem fs = FileSystem.get(conf);
+      // check whether the current user is the same one with the owner of hbase root
+      String currentUserName = UserGroupInformation.getCurrentUser().getShortUserName();
+      FileStatus[] hbaseRootFileStat = fs.listStatus(new Path(conf.get(HConstants.HBASE_DIR)));
+      if (hbaseRootFileStat.length > 0) {
+        String owner = hbaseRootFileStat[0].getOwner();
+        if (!owner.equals(currentUserName)) {
+          String errorMsg = "The current user[" + currentUserName
+              + "] doesn't have hbase root credentials."
+              + " If this job fails due to an inability to read HBase's internal directories, "
+              + "you will need to rerun as a user with sufficient permissions. The HBase superuser "
+              + "is a safe choice.";
+          LOG.warn(errorMsg);
```

Comment:
Thanks for only a WARN and not a fatal error. I do not know how MOB works with snapshot but can one leverage     [HBASE-18659 - "Use HDFS ACL to give user the ability to read snapshot directly on HDFS"](https://issues.apache.org/jira/browse/HBASE-18659) for getting all that is needed here?

Reply:
Maybe?  If the ACL syncing tries to keep the needed permissions in place for all active mob files, archived files, and files referenced via hlinks, but if it does then it would work.

wrt snapshots, currently this report isn't made to examine a snapshot, but in principal we should be able to reuse the same logic (and I guess add a constraint that every referenced hfile should also be in the snapshot manifest).

Reply:
Cool; was wondering and wanted to plant the seed too. I have no use case but appreciate the idea of HBASE-18659 as a way to help tenants not need `hbase` user privileges on HDFS.